src/common/health.rs

```diff
@@ -0,0 +1,359 @@
+        }
+
+        self.is_ready.store(true, atomic::Ordering::Relaxed);
+        self.notify_ready.notify_waiters();
+    }
+
+    /// Set ready to `true` if there is only one node in the cluster.
+    fn check_shortcut(&self) -> bool {
+        let peer_address_by_id = self.consensus_state.peer_address_by_id();
+
+        if peer_address_by_id.len() <= 1 {
+            self.is_ready.store(true, atomic::Ordering::Relaxed);
+            self.notify_ready.notify_waiters();
+            true
+        } else {
+            false
+        }
+    }
+
+    pub async fn get_largest_commit_in_cluster(&self) -> Option<u64> {
+        let this_peer_id = self.toc.this_peer_id;
+
+        let transport_channel_pool = self.toc.get_channel_service().channel_pool.clone();
+        let peer_address_by_id = self.consensus_state.peer_address_by_id();
+
+        let requests: FuturesUnordered<_> = peer_address_by_id
+            .iter()
+            .filter_map(|(&peer_id, uri)| {
+                if peer_id != this_peer_id {
+                    Some(uri)
+                } else {
+                    None
+                }
+            })
+            .map(|uri| get_commit_index(&transport_channel_pool, uri))
+            .collect();
+
+        // Example:
+        // Total nodes: 2
+        // Required: 2 / 2 = 1
+        //
+        // Total nodes: 3
+        // Required: 3 / 2 = 1
+        //
+        // Total nodes: 4
+        // Required: 4 / 2 = 2
+        //
+        // Total nodes: 5
+        // Required: 5 / 2 = 2
+        let required_commit_indices_count = peer_address_by_id.len() / 2;
+
+        let mut requests = requests
+            .inspect_err(|err| log::error!("GetCommitIndex request failed: {err}"))
+            .filter_map(|res| future::ready(res.ok()));
+
+        let commit_indices: Vec<_> = (&mut requests)
+            .take(required_commit_indices_count)
+            .collect()
+            .await;
+
+        if commit_indices.len() < required_commit_indices_count {
+            log::warn!("More than 50% of cluster is not responding to commit index requests");
+            return None;
+        }
+
+        let cluster_commit_index = commit_indices
+            .into_iter()
+            .map(|resp| resp.into_inner().commit)
+            .max()
+            .unwrap_or(0);
+
+        Some(cluster_commit_index as _)
+    }
+
+    fn commit_index(&self) -> u64 {
+        // TODO: Blocking call in async context!?
+        self.consensus_state
+            .persistent
+            .read()
+            .state
+            .hard_state
+            .commit
+    }
+
+    async fn check_ready(&mut self) -> bool {
+        for _ in 0..10 {
+            match self.check_ready.recv().await {
+                Ok(()) => return true,
+                Err(broadcast::error::RecvError::Closed) => return false,
+                Err(broadcast::error::RecvError::Lagged(_)) => (),
+            }
+        }
```

Comment:
Is 10 just an arbitrary value? Wouldn't a single retry not be enough with `Lagged(_)`? I'm unsure.

Reply:
I only found `Notify` channel today, and I think it would be easier to use `CancellationToken` + `Notify` than a single `broadcast` channel.

Reply:
What about a CondVar to prevent this situation completely? Not sure how well that works in async context though.

Reply:
Last time we had problems with `CondVar` in async, so I would rather use tokio here

Reply:
Replaced `broadcast` with `Notify` (and `CancellationToken`).